Add FjORD baseline

stevelaskaridis opened the pull request:
<!--
Thank you for opening a pull request (PR)!

Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

### Description

Implements [FjORD](https://openreview.net/forum?id=4fLr7H5D_eT) baseline.

### Related issues/PRs

See issue #2035.

## Proposal

Implementation of _FjORD: Fair and Accurate Federated Learning under heterogeneous targets with Ordered Dropout_ (published at NeurIPS'21) for Flower's Summer of Reproducibility.

### Explanation

The current PR implements FjORD baseline for summer of reproducibility. Specifically, we provide code for the ordered dropout implementation and replicate the results of Figure 5 of the paper.

### Checklist

- [x] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Any other comments?

<!--
Please be aware that it may take some time until the maintainers can review the PR.
Smaller PRs with good descriptions can be considered much more easily.

If you have an urgent request or question, please use the Flower Slack:

    https://flower.dev/join-slack/ (channel: #contributions)

Thank you for contributing to Flower!
-->

stevelaskaridis commented:
cc @SamuelHorvath

jafermarq commented:
Hi @stevelaskaridis, @SamuelHorvath

I started looking into the code but it is not following the expected file structure. The easiest way imo would be to, run the script in `<flower-repo>/baselines` that automatically creates the file structure from the [template_baseline]() contents. It's easy! just run:

```
./dev/create-baseline.sh fjord
```

from the aforementioned directory. Instructions [shown here](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline). This is the procedure all other baselines have follow. Also, part of the tests check for that particular file structure to be followed (you can add more files if needed). There is nothing super special about what the script above does, but it's desirable for all baselines to follow a similar structure.

Please ping me when I should look into the code again.

stevelaskaridis commented:
Hey @jafermarq, thanks for the feedback. We hadn't realise that the template was to be that strictly followed. In its current form, our baseline implementation should now be closer to what you were looking for.

Tests-wise, I am not sure the current failures are something on our side. But please let me know if it is something that we can fix.

jafermarq commented on baselines/fjord/fjord/main.py:
```diff
@@ -0,0 +1,288 @@
+        num_clients=NUM_CLIENTS,
+        config=fl.server.ServerConfig(num_rounds=args.num_rounds),
+        strategy=strategy,
+        client_resources=client_resources,
```
Let's pass this directly without parsing, it could be that the majority of the work (i.e. client side of things) is done on a different node than where the server lives.

jafermarq commented on baselines/fjord/fjord/main.py:
```diff
@@ -0,0 +1,288 @@
+            "loglevel": args.loglevel,
+            "logfile": args.logfile,
+        }
+        return FjORDClient(
```
Let's let the clients "choose" the device that the VirtualClientEngine exposes to them. No need to manually pass `device` to your `get_client_fn()`

jafermarq commented on baselines/fjord/README.md:
````diff
@@ -0,0 +1,108 @@
+### Through regular pip
+
+```bash
+cd fjord
+pip install -r requirements.txt
+pip install -e .
````
Updated assuming you moved those files up one level. I found `pip install -e . ` to be conflicting with Poetry's `pyproject.toml`

stevelaskaridis replied:
Resolved with [071af5e](https://github.com/adap/flower/pull/2431/commits/071af5e283f1a17e2ca582502ea7782c16a1ccd3)

jafermarq commented on baselines/fjord/fjord/conf/config.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+hydra:
+  run:
+    dir: ./runs/${now:%Y-%m-%d}:${now:%H-%M-%S}
```
I wonder if we need this. Hydra by defaults creates an experiment for your directory following the structure: `outputs/date/time`. E.g. `outputs/Y-M-D/h-m-s/`

stevelaskaridis replied:
I would prefer the flat structure run directories that is alphanumerically sorted by the filesystem.

jafermarq commented on baselines/fjord/fjord/conf/config.yaml:
```diff
@@ -0,0 +1,8 @@
+  run:
+    dir: ./runs/${now:%Y-%m-%d}:${now:%H-%M-%S}
+
+defaults:
+  - train_mode/fjord
+  - override hydra/job_logging: disabled
```
```suggestion
loglevel: info
logfile: run.log

manual_seed: 123
model: resnet18
dataset: cifar10
num_clients: 100
data_path: "./data"
num_workers: 4
evaluate_every: 10

cuda: true
batch_size: 32
lr: 0.1
lr_scheduler: static
optimiser: sgd
momentum: 0
nesterov: false
weight_decay: 1e-4

sampled_clients: 10
min_fit_clients: 2
client_selection: random  # or balanced
num_rounds: 500
local_epochs: 1
strategy: fjord_fedavg
client_resources:
  num_cpus: 1
  num_gpus: 0.2
knowledge_distillation: ???
p_s:
  - 0.2
  - 0.4
  - 0.6
  - 0.8
  - 1.0
client_tier_allocation: uniform
```

jafermarq replied:
Unless you are thinking on adding more configs or variation between them, I think it's better to keep things simple. The above will run `fjord`, changing `knowledge_distillation` to true does `fjord_kd`.

Having it like this defined, simplifies usage i think. And for instance it allows your `run.sh` to be compactly executed as a single command:
```
bash
python -m fjord.main --multirun knowledge_distillation=true,false manual_seed=123,124,125
```
Which will create a dir structure like the below (with one sub dir per run):

```
(fjord-py3.10) ➜  fjord git:(fjord-flwr) ✗ ls multirun/2023-10-02/19-20-09
0  1  2  3  4  5  multirun.yaml
```

stevelaskaridis replied:
Extensibility was the rationale behind this design.
I know that hydra can be intricate a times, but I think we don't hit this yet with this structure. 

Unless it's a hard requirement from your side, I would keep the structure as is. Let me know.

jafermarq reviewed (commented):
Hi @stevelaskaridis, @SamuelHorvath

Thanks for making those changes. I was able to run the code, so that's great!! 

Pleas find quite a few suggestions below (most of them are very minor and you can, if happy with it, just _commit_ the suggestion. In summary these changes propose to:
* Add the remaining of the content to `pyproject.toml` (so it matches that in the template)
* Execute the code from the top-level dir and treat `fjord` as a module -- this essentially requires making small adjustments to the `imports`. Some import paths could be shortened if `od/` and `utils/` are outside their own `fjord` sub dir. You'll find them all suggested below. The idea is for all baselines to run similar to:
```bash
#after poetry shell
python -m fjord.main <options-to-override-config-settings>
```
* To keep things aligned, could you move your `setup.py` and `requirements.txt` to the same directory as the readme?.
* I suggested a simplification to the configs. It doesn't introduces changes. just keeps them a bit simpler (simpler to override from CLI too)
* Suggested changes to the readme. just to make the env setup a bit more self-contained (uniformity across baselines is ideal)

I applied on my side all the above changes to the code and configs, things were running fine. 

It's great having a notebook! Could you move it to a directory called `docs` and placed at the same level as the readme (just for consistency with other baselines). Also, would it be possible to add the images to the readme?

Pleas ping me when you'd like me to take another look and happy to discuss about the suggested changes further.

stevelaskaridis commented on baselines/fjord/README.md:
```diff
@@ -0,0 +1,108 @@
+cd fjord
+poetry run python main.py  # without knowledge distillation
+# or
+poetry run python main.py +train_mode=fjord_kd  # with knowledge distillation
```
Resolved with [071af5e](https://github.com/adap/flower/pull/2431/commits/071af5e283f1a17e2ca582502ea7782c16a1ccd3)

stevelaskaridis commented on baselines/fjord/README.md:
```diff
@@ -0,0 +1,108 @@
+cd fjord
+python main.py  # without knowledge distillation
+# or
+python main.py +train_mode=fjord_kd  # with knowledge distillation
```
Resolved with [071af5e](https://github.com/adap/flower/pull/2431/commits/071af5e283f1a17e2ca582502ea7782c16a1ccd3)

stevelaskaridis commented:
Hi @jafermarq, 

Thanks for the feedback. Please find below the requested and applied changes as a checklist:

* [x] Add the remaining of the content to pyproject.toml (so it matches that in the template)
* [x] Execute the code from the top-level dir and treat fjord as a module
* [x] To keep things aligned, could you move your setup.py and requirements.txt to the same directory as the readme?.
* [ ] Suggested simplification to the configs. It doesn't introduces changes. just keeps them a bit simpler (simpler to override from CLI too)
* [x] Suggested changes to the readme. just to make the env setup a bit more self-contained (uniformity across baselines is ideal)
* [x] Move notebook to a directory called docs and placed at the same level as the readme (just for consistency with other baselines).
* [x] Also, would it be possible to add the images to the readme in a directory named _static?

jafermarq commented on baselines/fjord/README.md:
````diff
@@ -0,0 +1,119 @@
+### Through regular pip
+
+```bash
+cd fjord
````
```suggestion
```

stevelaskaridis replied:
Fixed with 57817c9

jafermarq reviewed (commented):
Hi @stevelaskaridis,

Thank you for applying those suggestions. I have listed a few more below. Probably the main two are:
* fix import in `utils/utils.py`.
* Install `fjord` using `python setup.py install` (else conflict with poetry's `.toml`).

Finally, if you run `./dev/format-baseline.sh fjord` (while sourced in your poetry env) you'll see several warning/errors (mostly in the form of `D102 Missing docstring in public method`) that are easy to fix for the most part. Then you'll need to run the tests `./dev/test-baseline.sh fjord`. These two are needed before `fjord` can be merged. Happy to help if some are harder to fix.

Please ping me once you had a chance to look into the above. Happy to discuss these further.

stevelaskaridis commented on baselines/fjord/fjord/models.py:
```diff
@@ -0,0 +1,266 @@
+    loss = 0.0
+    samples = 0
+    for _ in range(epochs):
+        for images, labels in tqdm(trainloader):
```
Fixed with ed9fd6d

stevelaskaridis commented on baselines/fjord/fjord/utils/utils.py:
```diff
@@ -0,0 +1,48 @@
+import torch
+from torch.nn import Module
+
+from utils.logger import Logger
```
Fixed with 4dcaf08

stevelaskaridis commented on baselines/fjord/fjord/conf/config.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+hydra:
+  run:
+    dir: ../runs/${now:%Y-%m-%d}:${now:%H-%M-%S}
```
Fixed with 93783f6

stevelaskaridis commented:
@jafermarq, thanks again for the valuable feedback. I have now:

* [x] Fixed the import in `utils/utils.py`
* [x] Amended the instructions to use setup.py
* [x] Removed tqdm from the dataloader 
* [x] Updated ray version
* [x] Changed code style to be in line with the tests.
* [x] Fixed typing issues to be in line with the tests.

A couple of issues:
- When initially running the mypy tests, it was having an [issue](https://github.com/python/mypy/issues/13639) with the numpy typing, so had to bump up the required version (you may want to update your instructions)
- I am currently having the issue of translating `FitRes` to the internal types we are using (where baselines/test fails). Not sure how to proceed on that one.

jafermarq commented:
Hi @stevelaskaridis,

Thanks for adding those changes. I tried running the code but it although it starts, something crashes (w/o error message) when doing the aggregation. The culprit is something added in commit [a0ff259](https://github.com/adap/flower/pull/2431/commits/a0ff2596aa83168832454391b300d37bad49263a)

jafermarq commented:
@stevelaskaridis , no worries. I made the following changes:
- disable type check in `FjORDFedAVG`'s `aggregate_fit()`.
- I had to rename your directory back to `fjord` (instead of `FjORD`). Without this the tests for `pylint` will fail because how they are currently setup. (directory and module name must match). 
- With the above solved, `pylint` test can start. Quite a few issues are flagged, i fixed ~50% of them. Most of the remaining one are of the form `C0103: Variable name "p" doesn't conform to snake_case naming style (invalid-name)`. 

(After the above i re-run `poetry run python -m fjord.main ` and the results came as expected, so I believe nothing broke)

Let me know once you have some time to check those. Please ping me if you encounter some issues. Happy to help

stevelaskaridis commented:
Hi @jafermarq,

Thanks for the previous changes. I have either solved or omitted the remaining ones. Hope this now works as intended.

A strange issue is that some pylint issues I could only see over github tests (despite using the poetry setup). 

The current structure tests will not pass as we don't need a `dataset_preparation.py` or `utils.py` file. Let me know how you would like to proceed.

jafermarq commented:
@stevelaskaridis nothing to be done. Thanks! I'll run the experiments once and all should be good to merge 👍
